### PR TITLE
Debug Panel: Introduce schema types

### PR DESCRIPTION
**Pull request by jackw**
**What is this feature?**

Schematizing debug panel

**Why do we need this feature?**

It's schema day!

**Who is this feature for?**

Schema lovers

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/plugins/panel/debug/panelcfg.gen.ts [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+//     public/app/plugins/gen.go
+// Using jennies:
+//     TSTypesJenny
+//     PluginTSTypesJenny
+//
+// Run 'make gen-cue' from repository root to regenerate.
+
+export const PanelCfgModelVersion = Object.freeze([0, 0]);
+
+export type UpdateConfig = {
```
Comment (jackw) [inline public/app/plugins/panel/debug/panelcfg.gen.ts:13]: @sdboyer we weren't quite sure how to approach this. The handwritten type definition looked like:

```
export type UpdateConfig = {
  [K in keyof UpdateCounters]: boolean;
};
```

Is there any support for `keyof` in cuetsy?

Comment (sdboyer) [reply public/app/plugins/panel/debug/panelcfg.gen.ts:13]: It's definitely possible to construct the equivalent rule in CUE, but it'd be pretty weird to teach cuetsy how to actually produce that output.

There's probably a happy medium here though where you put the expected logic into the CUE type, then use a handwritten veneer in TS to get what you want there

File: public/app/plugins/panel/debug/DebugPanel.tsx [hunk 2]
```diff
@@ -6,7 +6,7 @@ import { CursorView } from './CursorView';
 import { EventBusLoggerPanel } from './EventBusLogger';
 import { RenderInfoViewer } from './RenderInfoViewer';
 import { StateView } from './StateView';
-import { DebugPanelOptions, DebugMode } from './types';
+import { DebugPanelOptions, DebugMode } from './panelcfg.gen';
```
Comment (Clarity-89) [inline public/app/plugins/panel/debug/DebugPanel.tsx:9]: If you [re-export the generated types](https://github.com/grafana/grafana/pull/62121/files#diff-02e7ec2462e8170844513cd3d1ffc8d1247586c62e68af891aef0ae0bfaa748fR5) from `types.ts` instead, you won't need to update them in other places.

File: public/app/plugins/panel/debug/panelcfg.cue [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import (
+	// "github.com/grafana/grafana/packages/grafana-schema/src/common"
+	"github.com/grafana/grafana/pkg/plugins/pfs"
+)
+
+// This file (with its sibling .cue files) implements pfs.GrafanaPlugin
+pfs.GrafanaPlugin
+
+composableKinds: PanelCfg: {
+	maturity: "experimental"
```
Comment (Clarity-89) [inline public/app/plugins/panel/debug/panelcfg.cue:26]: I think before we can mark this as experimental, all the fields below should be documented.

File: public/app/plugins/panel/debug/panelcfg.cue [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+						UpdateConfig: {
+							render:        bool
+							dataChanged:   bool
+							schemaChanged: bool
+						} @cuetsy(kind="type")
+
+						UpdateCounters: {
+							render:        number
+							dataChanged:   number
+							schemaChanged: number
```
Comment (sdboyer) [inline public/app/plugins/panel/debug/panelcfg.cue:38]: Need to use a more specific numeric type, see [this](https://docs.google.com/document/d/1gPzSCQn-0Qg86PYpPcR-l_kTQ4Ba1wgnNLZFKPSe8yQ/edit#heading=h.cxy1gix5vuop)

File: public/app/plugins/panel/debug/panelcfg.gen.ts (hunk 1 above)
Comment (sdboyer) [reply public/app/plugins/panel/debug/panelcfg.gen.ts:13]: i pushed a commit that turns this into a derived thing

File: public/app/plugins/panel/debug/panelcfg.gen.ts [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+
+export enum DebugMode {
+  Cursor = 'cursor',
+  Events = 'events',
+  Render = 'render',
+  State = 'State',
+  ThrowError = 'ThrowError',
+}
+
+export interface DebugPanelOptions {
```
Comment (sdboyer) [inline public/app/plugins/panel/debug/panelcfg.gen.ts]: This has to be named `PanelOptions`, which is something we need to enforce in pfs. small bug in the way of letting us do that right now, but will have it quite soon

**Review by sdboyer — DISMISSED** [review]

File: public/app/plugins/panel/debug/DebugPanel.tsx (hunk 2 above)
Comment (ryantxu) [reply public/app/plugins/panel/debug/DebugPanel.tsx:9]: I think this is better -- simple enough to just update everywhere

**Review by ryantxu — APPROVED** [review]
This looks good!

File: public/app/plugins/panel/debug/panelcfg.cue [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+	maturity: "experimental"
+
+	lineage: {
+		seqs: [
+			{
+				schemas: [
+					{
+						UpdateConfig: {for k, _ in UpdateCounters {"\(k)": bool}} @cuetsy(kind="type")
+
+						UpdateCounters: {
```
Comment (ryantxu) [inline public/app/plugins/panel/debug/panelcfg.cue:35]: I think this is good -- but slightly weird and perhaps worth pointing out that `UpdateCounters` is not part of saved schema contract (it is only used at runtime, and not within the options)

Comment (jackw) [reply public/app/plugins/panel/debug/panelcfg.cue:35]: Hmm... any better if we only define `UpdateConfig` in this cue file then import it in a TS file and do the following?

```ts
export type UpdateCounters = {
  [K in keyof UpdateConfig]: number;
};
```

**Review by ryantxu — APPROVED** [review]
nice -- happy we are keeping runtime properties out of the schema :)

**Pull request merged**